### PR TITLE
Add personalized configuration parameters for each metastore.

### PR DESCRIPTION
**Pull request by yangyuxia**
### :pencil: Description
Different hive Metastores may have different parameter values for the same parameter. If multiple federated Metastores are configured in waggle dance, you can configure their own parameter values for each metastore.
For example, when hivemetastore enabled the kerberos authentication, the value of the "hive.metastore.kerberos.principal", some metastore is configured to "hive/_HOST@Realm", but some metastore is configured as "hive/{clusterName}@Realm", so you need to configure personalized configuration parameters for each metastore. 

### :link: Related Issues

### CODE REVIEW

**Comment by yangyuxia** [comment]
cat waggle-dance-server.yml

verbose: true
port: 48869
database-resolution: PREFIXED
yaml-storage:
 &nbsp;overwrite-config-on-shutdown: false
logging:
 &nbsp;config: file:${WAGGLE_DANCE_HOME}/conf/log4j2.xml
configuration-properties:
 &nbsp;hadoop.security.authentication: KERBEROS
 &nbsp;hadoop.kerberos.keytab.login.autorenewal.enabled: true
 &nbsp;hadoop.proxyuser.hive.users: '*'
 &nbsp;hadoop.proxyuser.hive.groups: '*'
 &nbsp;hadoop.proxyuser.hive.hosts: '*'
 &nbsp;hive.metastore.sasl.enabled: true
 &nbsp;hive.metastore.kerberos.principal: hive/_HOST@HADOOP.COM
 &nbsp;hive.metastore.kerberos.keytab.file: ${WAGGLE_DANCE_HOME}/conf/wd.keytab
 &nbsp;hive.cluster.delegation.token.store.class: org.apache.hadoop.hive.thrift.ZooKeeperTokenStore
 &nbsp;hive.cluster.delegation.token.store.zookeeper.connectString: host-zk1:2181,host-zk2:2181,host-zk3:2181
 &nbsp;hive.cluster.delegation.token.store.zookeeper.znode: /hive/cluster/delegationwg
 &nbsp;hive.cluster.delegation.token.store.zookeeper.acl: "world:anyone:rwcda"
 &nbsp;hive.server2.authentication: KERBEROS
 &nbsp;hive.server2.authentication.kerberos.keytab: ${WAGGLE_DANCE_HOME}/conf/wd.keytab
 &nbsp;hive.server2.authentication.kerberos.principal: hive/_HOST@HADOOP.COM
 &nbsp;hive.server2.authentication.client.kerberos.principal: hive/_HOST@HADOOP.COM
 &nbsp;wd-metastore:
 &nbsp;&nbsp;hms1:
 &nbsp;&nbsp;&nbsp;&nbsp;hive.metastore.kerberos.principal: hive/_HOST@HADOOP.COM
 &nbsp;&nbsp;hms2:
 &nbsp;&nbsp;&nbsp;&nbsp;hive.metastore.kerberos.principal: hive/test2@HADOOP.COM

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactory.java [hunk 1]
```diff
@@ -45,7 +43,21 @@ public class CloseableThriftHiveMetastoreIfaceClientFactory {
   public CloseableThriftHiveMetastoreIface newInstance(AbstractMetaStore metaStore) {
     Map<String, String> properties = new HashMap<>();
     if (waggleDanceConfiguration.getConfigurationProperties() != null) {
-      properties.putAll(waggleDanceConfiguration.getConfigurationProperties());
+//      properties.putAll(waggleDanceConfiguration.getConfigurationProperties());
+      Map<String, String> serverConfigMap=waggleDanceConfiguration.getConfigurationProperties();
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactory.java:47]: I suggest we move these properties to the AbstractMetastore class and make it part of the waggle-dance-federation.yml and this code can become much simpler like:

```
properties.putAll(waggleDanceConfiguration.getConfigurationProperties());
//override per metastore
properties.putAll(metastore.getConfigurationProperties());
```

**Review by patduin — CHANGES_REQUESTED** [review]
Thank you please have a look and my comment.

**Comment by yangyuxia** [comment]
primary-meta-store:
  database-prefix: ''
  name: hms1
  remote-meta-store-uris: thrift://host-hms1:9083
  access-control-type: READ_AND_WRITE_AND_CREATE
  impersonation-enabled: false
  configuration-properties:
    hive.metastore.kerberos.principal: hive/hms@HADOOP.COM
    hive.server2.enable.doAs: true

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactory.java [hunk 2]
```diff
@@ -47,6 +47,8 @@ public CloseableThriftHiveMetastoreIface newInstance(AbstractMetaStore metaStore
     if (waggleDanceConfiguration.getConfigurationProperties() != null) {
       properties.putAll(waggleDanceConfiguration.getConfigurationProperties());
     }
+    //override per metastore
+    properties.putAll(metaStore.getConfigurationProperties());
```
Comment (patduin) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactory.java:51]: Could you add a junit test please?

Comment (yangyuxia) [reply waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactory.java]: added

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java [hunk 3]
```diff
@@ -51,15 +51,15 @@ public abstract class AbstractMetaStore {
   private List<String> writableDatabaseWhitelist;
   private List<String> mappedDatabases;
   private @Valid List<MappedTables> mappedTables;
-  private Map<String, String> databaseNameMapping = Collections.emptyMap();
+  private transient Map<String, String> databaseNameMapping = Collections.emptyMap();
```
Comment (patduin) [inline waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java:54]: why this had to change?

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java [hunk 4]
```diff
@@ -243,4 +244,12 @@ public String toString() {
         .toString();
   }
 
+  public Map<String, String> getConfigurationProperties() {
```
Comment (jmnunezizu) [inline waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java:247]: Please move these two methods below the `setDatabaseNameMapping` method to keep them in sync with the property definitions. Normally, `toString` (if defined) is the last method in a class.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactory.java [hunk 5]
```diff
@@ -47,6 +47,10 @@ public CloseableThriftHiveMetastoreIface newInstance(AbstractMetaStore metaStore
     if (waggleDanceConfiguration.getConfigurationProperties() != null) {
       properties.putAll(waggleDanceConfiguration.getConfigurationProperties());
     }
+    //override per metastore
```
Comment (jmnunezizu) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactory.java:50]: The comment isn't needed. 
nit: there's a space missing between `if` and `(`. You might also want to leave an empty line before the `return` statement to improve readability.

File: waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactoryTest.java [hunk 6]
```diff
@@ -68,8 +70,9 @@ public void setUp() {
   @Test
   public void defaultFactory() {
     ArgumentCaptor<HiveConf> hiveConfCaptor = ArgumentCaptor.forClass(HiveConf.class);
-
-    factory.newInstance(newFederatedInstance("fed1", THRIFT_URI));
+    FederatedMetaStore fed1 = newFederatedInstance("fed1", THRIFT_URI);
+    fed1.setConfigurationProperties(Collections.singletonMap(ConfVars.METASTORE_KERBEROS_PRINCIPAL.varname,"hive/_HOST@HADOOP.COM"));
```
Comment (jmnunezizu) [inline waggle-dance-core/src/test/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIfaceClientFactoryTest.java:74]: Nit: there's a space missing between the key and the value.

**Review by jmnunezizu — CHANGES_REQUESTED** [review]

**Comment by jmnunezizu** [comment]
Thanks for the contribution. Please update the readme explaining how this works and how it can be utilised. Thanks.

**Review by jmnunezizu — APPROVED** [review]
Thanks!

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java [hunk 7]
```diff
@@ -51,15 +51,15 @@ public abstract class AbstractMetaStore {
...
+  private transient Map<String, String> databaseNameMapping = Collections.emptyMap();
   private @NotBlank String name;
   private @NotBlank String remoteMetaStoreUris;
   private @Valid MetastoreTunnel metastoreTunnel;
   private @NotNull AccessControlType accessControlType = AccessControlType.READ_ONLY;
   private transient @JsonProperty @NotNull MetaStoreStatus status = MetaStoreStatus.UNKNOWN;
   private long latency = 0;
   private transient @JsonIgnore HashBiMap<String, String> databaseNameBiMapping = HashBiMap.create();
-  private Map<String, String> configurationProperties = Collections.emptyMap();
+  private Map<String, String> configurationProperties;// = Collections.emptyMap();
```
[the remark this answers is not in the document]
Comment (patduin) [reply waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java:62]: @yangyuxia if you could revert or explain why this change was made I'm happy to approver once that's done,thanks!

File: waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java (hunk 3 above)
Comment (patduin) [reply waggle-dance-api/src/main/java/com/hotels/bdp/waggledance/api/model/AbstractMetaStore.java:54]: @yangyuxia if you could revert or explain why this change was made I'm happy to approver once that's done,thanks!

**Review by patduin — APPROVED** [review]
Thank you

**Pull request merged**